docs/design/paginators.md

```diff
@@ -70,127 +69,113 @@ used in a model.
 
+Pagination by default will always generate an extension function that returns a `Flow` over the normal operation output type. 
 
-```kotlin
-/**
- * An asynchronous data stream. This type is an asynchronous version of [Iterable].
- */
-interface SdkAsyncIterable<out T> {
-   operator fun iterator(): SdkAsyncIterator<T>
-}
+An example of this is demonstrated below (codegen):
 
-/**
- * An asynchronous [Iterator].
- */
-interface SdkAsyncIterator<out T> {
-   /**
-    * Returns true if more results are expected
-    */
-   suspend operator fun hasNext(): Boolean
+```kotlin
+fun LambdaClient.paginateListFunctions(initialRequest: ListFunctionsRequest): Flow<ListFunctionsResponse> {
+   return flow {
+      var cursor: kotlin.String? = null
+      var isFirstPage: Boolean = true
+      val req = initialRequest.copy {
+         this.marker = cursor
+      }
 
-   /**
-    * Returns the results for the next page or null when no more results are available.
-    * @throws NoSuchElementException if [hasMorePages()] is false
-    */
-   suspend operator fun next(): T
+      while (isFirstPage || (cursor?.isNotEmpty() == true)) {
+         val result = this@paginateListFunctions.listFunctions(req)
+         isFirstPage = false
+         cursor = result.nextMarker
+         emit(result)
+      }
+   }
 }
+```
 
-inline fun<T, R> SdkAsyncIterable<T>.map(transform: (T) -> R): SdkAsyncIterable<R> = TODO("not-implemented")
-inline fun<T, R> SdkAsyncIterable<T>.mapNotNull(transform: (T) -> R?): SdkAsyncIterable<R> = TODO("not-implemented")
-inline fun<T, R> SdkAsyncIterable<Iterable<T>>.flatMap(transform: (T) -> Iterable<R>): SdkAsyncIterable<R> = TODO("not-implemented")
+The shape targeted by `inputToken` is mapped to the `cursor` field. Each iteration the `cursor` is updated with the output field targeted by the `outputToken`.
 
-fun<T> SdkAsyncIterable<Iterable<T>>.flatten(): SdkAsyncIterable<T> = TODO("not-implemented")
-fun <T> SdkAsyncIterable<T?>.filterNotNull(): SdkAsyncIterable<T> = TODO("not-implemented")
+### Pagination over Nested Item
 
-suspend fun<T> SdkAsyncIterable<T>.toList(): List<T> = TODO("not-implemented")
-suspend fun<T> SdkAsyncIterable<T>.toSet(): Set<T> = TODO("not-implemented")
+If `items` is specified in the model for a given operation, then a `Flow` transform function (extending the return type of it's parent) will be generated that 
+allows the nested item to be paginated over by producing another flow that works from the base response flow.
+The targeted type will provide iteration to the `item` element for the user automatically.
 
-/**
- * Terminal operator that collects the given iterable and calls [action] for each result
- */
-inline fun <T> SdkAsyncIterable<T>.collect(crossinline action: suspend (T) -> Unit):Unit = TODO()
+Here is an example that follows from the previous example (codegen):
+
+```kotlin
+fun Flow<ListFunctionsResponse>.onFunctionConfiguration(): Flow<FunctionConfiguration> =
```

Comment:
I don't love this name. `on` implies event handling to me. We could probably just name it after the name of `items` (and in the case of longer nesting use the leaf name)

```kotlin
fun Flow<ListFunctionsResponse>.functions(): Flow<FunctionConfiguration>
```

Reply:
I see your point about the overuse of `on`.  what about `with` or `over`?  Regarding the rest of the function name, I think that specifying the type that the customer will collect in the flow is helpful over the member name, but see your point about using the terminal member specified by items.  I could go either way.

Reply:
> I think that specifying the type that the customer will collect in the flow is helpful over the member name,

I could see that. I see it from the member name which is what customers interact with when they aren't paginating. The lambda example may be overly simplified but *if* it's even a little bit representative `functions` is clearer to me than the type itself. Probably good to hear from @ianbotsf or @kiiadi for their gut reaction here as well.

Reply:
I know this is anecdotal but as a consumer I also agree with @aajtodd. We already use the *items'* property name to access it, I believe it'll be natural to have the same as an extension function on the flow.

Also works better (?) in the case of DynamoDb queries because I'm not sure what name the initial proposal wouldve generated vs below:
```
fun Flow<QueryResponse>.items(): Flow<List<Map<String, AttributeValue>>?>
```

Reply:
Hi @efemoney , thanks for your feedback!  Yes in fact we received more feedback internally regarding this point in particular.  I've just updated the design and the current plan is to generate the mapping function to always be named `items()`.  Here is how the transform would look:

```kotlin
fun Flow<ScanResponse>.items(): Flow<Map<String, AttributeValue>>
```

and sample usage:

```kotlin
dynamoDbClient
    .paginateScan(ScanRequest { tableName = "some-table" })
    .items()
    .collect { item -> // Map<String, AttributeValue>
        println(item)
    }
```

Any thoughts on this?

Reply:
To clarify my suggestion was not to always name it `items()` but rather use the member name targeted by the `@paginated` trait `items` field. In this case it's confusing because `DynamoDb::Scan` actually targets a member named [Items](https://github.com/awslabs/aws-sdk-kotlin/blob/v0.9.5-beta/codegen/sdk/aws-models/dynamodb.2012-08-10.json#L7335). 

So to illustrate with a less confusing example, the `DynamoDb::ListTables` paginator has a [target member](https://github.com/awslabs/aws-sdk-kotlin/blob/v0.9.5-beta/codegen/sdk/aws-models/dynamodb.2012-08-10.json#L5170) of `TableNames` so using my suggestion it would result in an extension like:

```kotlin
fun Flow<ListTablesResponse>.tableNames(): Flow<...>
```

I think this makes the most sense as it has the best symmetry between regular (un-paginated) and paginated responses:

e.g.
**Un-paginated**
```kotlin
val resp = lambda.listFunctions()

resp.functions?.forEach { ... }
```

**Paginated**

```kotlin
lambda.listFunctionsPaginated()
    .functions()     // extension function for the member name targeted by `@paginated.items`
    .collect()
```

In the case of something like redshift which targets a nested member [DefaultClusterParameters.Parameters](https://github.com/awslabs/aws-sdk-kotlin/blob/v0.9.5-beta/codegen/sdk/aws-models/redshift.2012-12-01.json#L5589) we would use the leaf member name (so in this case just `parameters()`). This [appears](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/redshift/paginators/DescribeDefaultClusterParametersIterable.html) to be what Java V2 is doing as it turns out.

Reply:
I agree with @aajtodd here. Using `items` may be confusing because:
* It doesn't match with non-paginated call patterns (e.g., there is no `items` member on a regular result object)
* It may conflict with service-local semantics of an Item (e.g., DynamoDB Items won't be returned by a paginated `ListTables` paginated call despite having an `items` member)

Reply:
I think `items` (or some other consistent name) works better than using the member name. In my experience*, most people prefer consistency since once you're working with several different services any amount of inconsistency in interfaces becomes a major sore spot.

> It doesn't match with non-paginated call patterns (e.g., there is no `items` member on a regular result object)

I don't think this is a bad thing. A paginated operation *isn't* the a normal call, and shouldn't be easily confused with one.

> It may conflict with service-local semantics of an Item

I think this is an acceptable trade-off, but would something like `.iterator()` work better? I'm not sure how / if that term is used in the rust community, but to my mind it would be a good description of what the function is returning.

*For some context, this is mostly informed by my experience of working on the Python SDK and CLI for several years. We got a *lot* of complaints any time there were inconsistencies.

Reply:
> I think items (or some other consistent name) works better than using the member name. In my experience*, most people prefer consistency since once you're working with several different services any amount of inconsistency in interfaces becomes a major sore spot.

Agreed that inconsistency is bad but consistency is relative to something. I still think using the member name is the right choice as it provides consistency w.r.t to the singular response. Using the literal identifier `items()` provides consistency w.r.t other services/calls but it isn't descriptive at all as to what it means. Customers would have to look at the function signature and documentation to figure out what is being paginated. This is particularly true if the thing being paginated over is builtin like `String`. 

e.g. The DDB `ListTables` paginator paginates over `TableNames` which is a list of strings. These will be flattened so the resulting signature would look like:

```kotlin
// Returns a flow of table names
fun Flow<ListTablesResponse>.items(): Flow<String>
```

What does that `String` represent without looking at the documentation? I think `tableNames()` would be much clearer.

Reply:
> What does that `String` represent without looking at the documentation? I think `tableNames()` would be much clearer.

It represents the `Tables` from `ListTables`. The member name for the paginated response list is typically stuttered from the operation name.

> consistency w.r.t to the singular response

This is not something we want. The singular response is not the same thing. Confusing the two is already a problem in other languages that have a stronger distinction than this.

Reply:
> It represents the Tables from ListTables. 

Yes *I* understand what it is, my point was that the name `items()` tells _customers_ nothing about it without looking at the documentation.

> The member name for the paginated response list is typically stuttered from the operation name.

I'm not sure what you mean here exactly. The docs for the [paginated trait](https://awslabs.github.io/smithy/1.0/spec/core/behavior-traits.html#paginated-trait) state that `items` is:

> The path to an output member of the operation that contains the data that is being paginated across many responses. The named output member, if specified, MUST target a list or map.


---------

> This is not something we want. The singular response is not the same thing. Confusing the two is already a problem in other languages that have a stronger distinction than this.

I don't feel like we are having the same conversation here. The `items` targeted by the trait must target an output member name. When I say consistency w.r.t to the singular response that's what I'm referring to. If a customer interacts with the singular response it looks like:

```kotlin
val singular = lambda.listFunctions { }
singular.functions?.forEach { ... }
```

Paginated using the member name:

```kotlin
lambda.listFunctionsPaginated()
     .functions()
     .collect { ... }
```

This feels like the right consistency to me. It's clear at a glance what I'm trying to do and what is being paged over.

Reply:
why not both?

Reply:
> Yes I understand what it is, my point was that the name `items()` tells customers nothing about it without looking at the documentation.

That method invocation doesn't exist in a void, it must only follow after a statement that give the necessary context.

> >The member name for the paginated response list is typically stuttered from the operation name.

> I'm not sure what you mean here exactly.

Operation outputs must always be structures, so services teams must give some name to the list they're returning. In the majority of cases, the name of the property is a suffix of the name of the operation or something very close to that. E.g.:

Operation name `ListFoos` becomes output structure member `ListFoosResponse.foos` becomes operation invocation `client.listFoos().foos`.

It's stuttered because it's duplicated information, essentially. You can't avoid that in a singular invocation, but there's no need to preserve it if you're iterating. So you don't lose context going from `listFoosPaginated.foos()` to `listFoosPaginated.items()`.

> I don't feel like we are having the same conversation here.

There are three cases.

1. Single operation invocation:

```
val singular = lambda.listFunctions { }
singular.functions?.forEach { ... }
```

2. Paginating:

```
lambda.listFunctionsPaginated()
     .collect { ... }
```

3. Iterating:

```
lambda.listFunctionsPaginated()
     .items()
     .collect { ... }
```

I'm concerned with making all of these as distinct as possible on an operation level while making each case consistent across operations. 2 and 3 looking like each other is not great because the return types are different and the way you have to interact with them is very different. 1 and 3 looking like each other is not good because 1 is not paginated and confusion there can mean that not all items are looked at (if they thought they were paginating but were not) or that too many are looked at (if they didn't think they were paginating but were). These might sound like purely theoretical issues, but I have personally assisted customers with these issues in the past.

Reply:
Thank you for clarifying.

Reply:
In the design review we opted to go with the following, and will solicit feedback from customers once an implementation is available for them to test drive.  Either way the codegen changes are minimal.

> use the member name targeted by the @paginated trait items field